doc/explanation/package-management.md

```diff
@@ -0,0 +1,220 @@
+# How Package Management Works
+
+This document aims to give an explanation on how the new package management
+feature introduced in Dune works under the hood. It requires a bit of
+familiarity with how OPAM repositories work and how Dune builds packages. Thus
+it is aimed at people who want to understand how the feature works, not how it
+is used.
+
+For a tour on how to apply package management to a project refer to the
+{doc}`/tutorials/dune-package-management/index` tutorial.
+
+## Motivation
+
+A core part of modern programming is using other existing code to save time.
+The history of the OCaml package ecosystem is quite long with a lot of projects
+building upon each other over many years. A significant step forward was the
+creation of an OCaml Package Manager, OPAM, along with the establishment of a
+public package repository which made it a lot more feasible to share code
+between people and projects.
+
+Over time best practices have evolved, and while OPAM has incorporated some
+changes, due to its existing user base and constraints it couldn't adopt all
+the modern workflows.
+
+Dune package management attempts to take the parts of the OPAM ecosystem that
+have stood the test of time and couple them with a modern workflow. Some of the
+improvements include:
+
+* Automatic package repository updates
+* Easily reproducible dependencies
+* All package dependencies declared in a single file that is kept in-sync
+* Per-project dependencies
```

Comment:
I don't see what's new in that list from what opam can already do (apart from automatic package repository updates, which can be seen as quite bad from a reproducibility point of view if you don't have a lockfile). 

Instead of comparing opam vs. dune pkg, I would stress the design principle. A random list of stuff:
- making the OCaml environment setup trivial;
- promoting a lockfile-first approach to address reproducibility use cases;
- unifying the configuration files (and removing hidden global states managed by a CLI tool);
- improving cross-packages and vendoring workflows;
and much more.

Reply:
Here's a few principles that I've had in mind when designing this feature:

1. No global state visible to the user
2. All package configuration must be done via config file updates (dune-project or dune-workspace). No stateful commands such as `opam pin`.
3. "Automatic` package repository updates" also applies to pins. For example, we re-fetch branches & tags.
4. Package builds can only access packages which are listed as dependencies.
5. Users do not need to learn yet another file format to configure their workspace. Everything should be doable via workspaces.
6. Build plans are independently versioned so that they can be interpreted in the same way between different versions of dune

Reply:
I've updated the section to describe the features of the package management without explicit comparison with OPAM.